src/Compose/ConcurrentTransitionManagerContainer.cs

```
@@ -1,88 +1,88 @@
-			while (!managers.IsEmpty)
-				if (managers.TryTake(out manager))
-					restoredManagers.Add(Restore(Helper.MakeGenericType(serviceType).GetTypeInfo(), manager));
-			while (!restoredManagers.IsEmpty)
-				if (restoredManagers.TryTake(out manager))
-					managers.Add(manager);
-		}
-
-		private object Restore(TypeInfo helper, object manager)
-		{
-			helper.DeclaredMethods.Single(x => x.Name == "Restore").Invoke(null, new[] { manager });
-			return manager;
-		}
-
-		public void Snapshot()
-		{
-			foreach (var kvp in Managers)
-				Snapshot(kvp.Key, kvp.Value);
-		}
-
-		private void Snapshot(Type serviceType, ConcurrentBag<object> managers)
-		{
-			var snapshottedManagers = new ConcurrentBag<object>();
-			object manager = null;
-			while (!managers.IsEmpty)
-				if (managers.TryTake(out manager))
-					snapshottedManagers.Add(Snapshot(Helper.MakeGenericType(serviceType).GetTypeInfo(), manager));
-			while (!snapshottedManagers.IsEmpty)
-				if (snapshottedManagers.TryTake(out manager))
-					managers.Add(manager);
-		}
-
-		private object Snapshot(TypeInfo helper, object manager)
-		{
-			helper.DeclaredMethods.Single(x => x.Name == "Snapshot").Invoke(null, new[] { manager });
-			return manager;
-		}
-	}
-}
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Reflection;
+
+namespace Compose
+{
+	internal sealed class ConcurrentTransitionManagerContainer : TransitionManagerContainer
+	{
+		private readonly ConcurrentDictionary<Type, ConcurrentBag<object>> Managers
+			= new ConcurrentDictionary<Type, ConcurrentBag<object>>();
+		private static readonly Type Helper = typeof(GenericHelper<>);
+
+		private static class GenericHelper<T>
+		{
+			private static readonly TypeInfo Disposable = typeof(IDisposable).GetTypeInfo();
+			public static void Restore(object untypedManager)
+			{
+				var manager = untypedManager as DynamicRegister<T>;
+				if (Disposable.IsAssignableFrom(manager.CurrentService.GetType().GetTypeInfo()))
```

> ""but I think that we should still be checking anyway (everywhere, not just here) as we shouldn't be trusting the logic elsewhere. What if that logic changes?""

Either tests should fail, or we're missing tests?